Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -127,11 +129,23 @@ class Writer {
    *
    * @param value The value to write, represented as a uint64_t.
    */
-  void WriteVarint(uint64_t value);
+  Status WriteVarint(uint64_t value);
 
   pb_ostream_t stream_;
 };
 
+/**
+ * Returns an internal error status, or unconditionally dies in DEV mode via a
+ * failed assertion. This method should only be used for errors that strongly
+ * violate our underlying assumptions about the state of the system. (All nanopb
+ * serializing (but not deserializing) errors fall into this category, as our
+ * code should only be presenting valid objects for serialization.)
+ */
+Status getInternalError(const char* errmsg) {
```

Comment:
> We'd have to do that a lot. It bloats the code without adding much value. OTOH, it only "bloats" it by a single statement or two each time. Done.

Sorry, that's not what I meant. I meant handling it only at the *top-level*, that is, only the code that calls the public method of `Serializer` should do that, and `Writer` should just propagate the status.

Reply:
Ah, I see; yeah that makes sense, but moves the assertion further away from the problem location. The stack trace would be less useful if we waited until the top level caller before asserting.

Reply:
> The stack trace would be less useful if we waited until the top level caller before asserting.

Ah, that's reasonable (for serialization/deserialization, I don't find stack trace all that useful because it doesn't really say which field failed, but it's still better than nothing)